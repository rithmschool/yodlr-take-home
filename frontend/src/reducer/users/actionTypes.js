export const LOAD_USERS_TO_STORE = 'LOAD_USERS_TO_STORE';
export const ADD_USER_TO_STORE = 'ADD_USER_TO_STORE';
export const REMOVE_USER_FROM_STORE = 'REMOVE_USER_FROM_STORE';
export const UPDATE_USER_IN_STORE = 'UPDATE_USER_IN_STORE';