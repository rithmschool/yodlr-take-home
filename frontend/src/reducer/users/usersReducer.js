import {
  LOAD_USERS_TO_STORE, ADD_USER_TO_STORE, REMOVE_USER_FROM_STORE, UPDATE_USER_IN_STORE
} from './actionTypes';


function usersReducer(state = {}, action) {
  switch (action.type) {
    case LOAD_USERS_TO_STORE:
      const users = action.users;
      return users.reduce((table, user) => {
        table[user.id] = user;
        return table;
      }, {});

    case ADD_USER_TO_STORE:
      return { ...state, [action.user.id]: action.user };

    case REMOVE_USER_FROM_STORE:
      const data = { ...state };
      delete data[action.id];
      return data;

    case UPDATE_USER_IN_STORE:
      return { ...state, [action.user.id]: { ...action.user } };

    default:
      return state;
  }
}


export default usersReducer;