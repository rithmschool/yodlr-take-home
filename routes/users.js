var express = require('express');
var router = express.Router();
var _ = require('lodash');
var logger = require('../lib/logger');
var log = logger();
const { ensureAdmin } = require('../helpers/auth.js');

var users = require('../init_data.json').data;
var curId = _.size(users);

/* GET users listing. */
router.get('/', ensureAdmin, function(req, res) {
  return res.json(_.toArray(users));
});

/* Create a new user */
router.post('/', function(req, res) {
  var user = req.body;
  user.id = curId++;
  if (!user.state) {
    user.state = 'pending';
  }
  users[user.id] = user;
  log.info('Created user', user);
  return res.json(user);
});

/* Get a specific user by id */
router.get('/:id', ensureAdmin, function(req, res, next) {
  var user = users[req.params.id];
  if (!user) {
    return next();
  }
  return res.json(users[req.params.id]);
});

/* Delete a user by id */
router.delete('/:id', ensureAdmin, function(req, res) {
  var user = users[req.params.id];
  delete users[req.params.id];
  res.status(204);
  log.info('Deleted user', user);
  return res.json(user);
});

/* Update a user by id */
router.put('/:id', ensureAdmin, function(req, res, next) {
  var user = req.body;
  if (user.id != req.params.id.id) {
    return next(new Error('ID paramter does not match body'));
  }
  users[user.id] = user;
  log.info('Updating user', user);
  return res.json(user);
});


module.exports = router;
