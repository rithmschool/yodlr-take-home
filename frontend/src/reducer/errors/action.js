import { SET_ERROR, RESET_ERROR } from './actionTypes';


function setError(message) {
  return { type: SET_ERROR, message };
}

function resetError() {
  return { type: RESET_ERROR };
}


export { setError, resetError };