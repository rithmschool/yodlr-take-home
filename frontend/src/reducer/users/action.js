import {
  LOAD_USERS_TO_STORE, ADD_USER_TO_STORE, REMOVE_USER_FROM_STORE, UPDATE_USER_IN_STORE
} from './actionTypes';
import User from '../../api/User';
import { setError, resetError } from '../errors/action';


function storeUser(user) {
  return { type: ADD_USER_TO_STORE, user };
}

/**
 * Send post request to backend and add new user to redux store.
 * @param {Object} user - { email, firstName, lastName, state }
 */
function addUser(user) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      const newUser = await User.create(user);
      dispatch(storeUser(newUser));
    } catch (error) {
      console.error(error);
      setError(error.message);
    }
  }
}

function removeUserFromStore(id) {
  return { type: REMOVE_USER_FROM_STORE, id };
}

/**
 * Sends DELETE request to backend and remove user from redux store.
 * @param {Number} id 
 */
function removeUser(id) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      await User.remove(id);
      dispatch(removeUserFromStore(id));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}

function loadUsersToStore(users) {
  return { type: LOAD_USERS_TO_STORE, users };
}

/**
 * Sends GET request to backend and loads users to redux store.
 */
function loadUsers() {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      const users = await User.getAll();
      dispatch(loadUsersToStore(users));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}

function updateStoredUser(user) {
  return { type: UPDATE_USER_IN_STORE, user };
}

function updateUser(user) {
  return async function(dispatch) {
    dispatch(resetError());

    try {
      await User.update(user.id, user)
      dispatch(updateStoredUser(user));
    } catch (error) {
      console.error(error);
      dispatch(setError(error.message));
    }
  }
}


export { addUser, removeUser, updateUser, loadUsers };