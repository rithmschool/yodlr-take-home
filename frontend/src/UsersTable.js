import React from 'react';
import { useSelector, shallowEqual, useDispatch } from 'react-redux';
import { removeUser, updateUser } from './reducer/users/action';


const UsersTable = () => {
  const data = useSelector(state => state.users, shallowEqual);
  const users = Object.values(data).sort((a, b) => +a.id - b.id);
  const dispatch = useDispatch();

  const handleActivate = (user) => {
    dispatch(updateUser({ ...user, state: "active" }));
  }

  return (
    <div className="UsersTable card mt-5">
      <table className="table">
        <thead>
          <tr>
            <th scope="col">#</th>
            <th scope="col">First Name</th>
            <th scope="col">Last Name</th>
            <th scope="col">Email</th>
            <th scope="col">State</th>
            <th scope="col"></th>
            <th scope="col"></th>
          </tr>
        </thead>
        <tbody>
          {users.map(user => (
            <tr key={user.id}>
              <th scope="row">{user.id}</th>
              <td>{user.firstName}</td>
              <td>{user.lastName}</td>
              <td>{user.email}</td>
              <td>{user.state}</td>
              <td>
                {user.state === "pending" &&
                  <button className="btn btn-warning" onClick={() => handleActivate(user)} data-testid={`activateBtn-${user.id}`}>
                    Activate
                  </button>}
              </td>
              <td>
                <button className="btn text-danger" onClick={() => dispatch(removeUser(user.id))} data-testid={`removeBtn-${user.id}`}>
                  <i className="fa fa-times" aria-hidden="true"></i>
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  )
}


export default UsersTable;