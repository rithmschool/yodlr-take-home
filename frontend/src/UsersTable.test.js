import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import { getByTestId, render, fireEvent } from '@testing-library/react';
import UsersTable from './UsersTable';
import { removeUser, updateUser } from './reducer/users/action';


jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

jest.mock('./reducer/users/action', () => ({
  removeUser: jest.fn(),
  updateUser: jest.fn(),
}))

beforeEach(function() {
  useSelector.mockImplementation((selectorFn) =>
    selectorFn({
      users: {
        0: {
          id: 0, firstName: "test0", lastName: "user", state: "active"
        },
        1: {
          id: 1, firstName: "test1", lastName: "user", state: "pending"
        }
      }
    })
  );
});

// smoke test
it("renders without crashing", function() {
  render(<MemoryRouter><UsersTable /></MemoryRouter>);
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(<MemoryRouter><UsersTable /></MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});

it("Dispatches removeUser() when clicking x icon", function() {
  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);
  const { getByTestId } = render(<MemoryRouter><UsersTable /></MemoryRouter>);

  const removeBtn = getByTestId("removeBtn-0");
  fireEvent.click(removeBtn);

  expect(mockedDispatch).toHaveBeenCalledWith(removeUser(0));
});

it("Dispatches updateUser() when click on activate button", function() {
  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);
  const { getByTestId } = render(<MemoryRouter><UsersTable /></MemoryRouter>);

  const activateBtn = getByTestId("activateBtn-1");
  fireEvent.click(activateBtn);

  expect(mockedDispatch).toHaveBeenCalledWith(updateUser({}));
});