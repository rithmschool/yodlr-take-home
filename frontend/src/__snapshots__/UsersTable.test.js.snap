// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="UsersTable card mt-5"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            #
          </th>
          <th
            scope="col"
          >
            First Name
          </th>
          <th
            scope="col"
          >
            Last Name
          </th>
          <th
            scope="col"
          >
            Email
          </th>
          <th
            scope="col"
          >
            State
          </th>
          <th
            scope="col"
          />
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            scope="row"
          >
            0
          </th>
          <td>
            test0
          </td>
          <td>
            user
          </td>
          <td />
          <td>
            active
          </td>
          <td />
          <td>
            <button
              class="btn text-danger"
              data-testid="removeBtn-0"
            >
              <i
                aria-hidden="true"
                class="fa fa-times"
              />
            </button>
          </td>
        </tr>
        <tr>
          <th
            scope="row"
          >
            1
          </th>
          <td>
            test1
          </td>
          <td>
            user
          </td>
          <td />
          <td>
            pending
          </td>
          <td>
            <button
              class="btn btn-warning"
              data-testid="activateBtn-1"
            >
              Activate
            </button>
          </td>
          <td>
            <button
              class="btn text-danger"
              data-testid="removeBtn-1"
            >
              <i
                aria-hidden="true"
                class="fa fa-times"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
